cmd/changelog/cmd/generate.go

```diff
@@ -0,0 +1,300 @@
+	var maxVersion *semver.Version
+	var maxRef *plumbing.Reference
+	for {
+		tag, err := tags.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, nil, err
+		}
+		ver, err := semver.NewVersion(tag.Name().Short())
+		if err == nil &&
+			(maxVersion == nil || ver.GreaterThan(maxVersion)) {
+			maxVersion = ver
+			maxRef = tag
+		}
+	}
+	return maxVersion, maxRef, nil
+}
+
+// findNewCommits queries the git repository for new commits since the last tagged release.
+func findNewCommits(r *git.Repository, prevHash plumbing.Hash) ([]Commit, error) {
+	logs, err := r.Log(new(git.LogOptions))
+	if err != nil {
+		return nil, err
+	}
+	defer logs.Close()
+
+	var commits []Commit
+	for {
+		commit, err := logs.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if commit.Hash == prevHash {
+			break
+		}
+		c, err := parseCommit(commit)
+		if err != nil {
+			return nil, err
+		}
+		commits = append(commits, c)
+	}
+	return commits, nil
+}
+
+const (
+	Feature = "feat"
+	Fix     = "fix"
+)
+
+// organizeCommits filters commits into groups based on their type.
+func organizeCommits(commits []Commit) (breaking, features, fixes []Commit) {
+	breaking = make([]Commit, 0, len(commits))
+	features = make([]Commit, 0, len(commits))
+	fixes = make([]Commit, 0, len(commits))
+
+	for _, c := range commits {
+		if c.Breaking {
+			breaking = append(breaking, c)
+		}
+		switch c.Type {
+		case Feature:
+			features = append(features, c)
+		case Fix:
+			fixes = append(fixes, c)
+		default:
+			//Ignore commit
+		}
+	}
+	return
+}
+
+// ReleaseType corresponds the semantic version numbers.
+type ReleaseType int
+
+const (
+	Major ReleaseType = iota
+	Minor
+	Patch
+)
+
+// determineNextVersion computes the next semantic version based on the types of commits in the release.
+func determineNextVersion(prevVer *semver.Version, numBreaking, numFeatures int) *semver.Version {
+	typ := Patch
+	if numBreaking > 0 {
+		typ = Major
+		if prevVer.Major() == 0 {
+			// Downgrade breaking change to minor release for pre 1.0.0 releases
+			typ = Minor
+		}
+	} else if numFeatures > 0 {
+		typ = Minor
+	}
+	v := incVersion(typ, *prevVer)
+	return &v
+}
+
+// incVersion increments the semantic version according to the release type.
+func incVersion(typ ReleaseType, v semver.Version) semver.Version {
+	switch typ {
+	case Major:
+		return v.IncMajor()
+	case Minor:
+		return v.IncMinor()
+	case Patch:
+		return v.IncPatch()
+	default:
+		panic("unreachable")
+	}
+}
+
+// Commit represents the conventional commit understanding of a git commit.
+type Commit struct {
+	Type    string
+	Scope   string
+	Subject string
+	Body    string
+	Footer  string
+
+	Breaking bool
+
+	Hash      string
+	ShortHash string
+	Author    string
+}
+
+// headerPattern matches the type, scope and subject of a commit message header.
+var headerPattern = regexp.MustCompile(`^(\w+)(\(\w+\))?:(.*)$`)
+
+const breaking = "BREAKING CHANGE:"
```

Comment:
The spec says "A breaking change MUST consist of the uppercase text BREAKING CHANGE, followed by a colon and a space," but it's probably fine if we're more relaxed here and don't require the space.

Reply:
yeah I had the space initially, but since we can't validate those before merge, I figured we would use a more relaxed search since as noted in a commit a false negative is more expensive than a false positive.